designs/kindless-upgrades.md

```diff
@@ -0,0 +1,150 @@
+
+In order to increase our confidence on this change, we will leverage our experience to try to find any edge cases that this new architecture could introduce. We will create a set of automated tests that run continuously through different scenarios and registering failures. This will allow us to investigate and find edge cases where special handling could be useful to avoid intermittent issues.
+
+* Scenarios to test across providers:
+    * k8s upgrades 1.24 → 1.25 → 1.26 → 1.27
+        * External etcd
+        * Stacked etcd
+    * Scaling up
+        * External etcd
+        * Stacked etcd
+    * Single and multi-node CP
+* Bare-metal will need a special test for single node cluster.
+
+We decide to run this separately from the main E2E CI to shorten the feedback loop. These only run on PR merges and they sometimes take up to 5h to complete. Running a smaller set of tests (which should reduce the execution time) in a continuous loop should increase the number of data points we obtain in the same period of time.
+
+The tests stats need to reflect if the tests failed during the upgrade or not, which is what matters here. If it failed at create or delete, just mark the tests as invalid, but not necessarily failed, so we don't invest time investigating the flake.
+
+It's important to remember that speed here is key, the sooner we get this running, the more data we will have. We should favor speed over perfection: use any tools at your disposal and avoid trying to automate everything.
+
+### Mitigating risks: a more robust reconciliation loop
+
+#### Separating etcd from CP changes
+
+Currently, the controller performs changes to the etcd cluster and CP all at once in one single step. This is no a problem from a workflow perspective, since the etcdadm cluster, CAPI cluster and KCP controllers orchestrate the operation in the right order. However, it makes recovering from an error more difficult since there are more changes to "rollback".
+
+Most of the time when an upgrade fails, it shows up in the first etcd node not coming up. By making the etcd changes a separate step in the upgrade process, we reduce the amount of changes to rollback in case of an issue. If the first etcd node doesn't come up, the etcd cluster spec can be manually changed to its previous version and the new machine deleted, leaving the cluster in the same state as before.
+
+This will require changing the way the etcdadm cluster, core CAPI cluster and KCP controllers interact, or at least make the current behavior optional. Right now, the system is designed to apply all the CP changes at once, so it orchestrates itself making that assumption. If this we don't change this, it will cause a double CP upgrade, one after etcd finishes the upgrade (when it updated the endpoints secret) and a second one after updating the KCP.
```

Comment:
I am a little confused how this is avoided in the new way, unless you are saying it doesn't until we actually change the reconciliation logic.

Reply:
Ah good point, I might need to rephrase this, jackson had a similar comment. Maybe you can help me find where the whole is.

Right now, we don't need to solve this problem, it's already solved. The 3 controllers coordinate themselves using a combination of annotations and objects to avoid double CP rollouts. But, the current orchestration only works if changes to etcdamdcluster and KCP are applies at the same time.

So, if we decide to execute this idea I propose here (spliting etcd from KCP in the eks-a controller), we will lose this "double rollout" protection. That's why I mention "It will require changing the way..." that this orchestration works.

Reply:
Oh I see okay, the section makes it seem like we are mitigating risk, but for this specific one we are adding a new concern, so maybe calling it out as a side effect or another action item might help? Not sure if you want to add another section below for specific points that need to be figured out still.

Reply:
I reworded this a bit, let me know if it helps.